lib/kinesis_client/stream/shard/lease.ex

```diff
@@ -149,74 +153,87 @@ defmodule KinesisClient.Stream.Shard.Lease do
-        Logger.debug(
-          "[kcl_ex] Failed to renew lease, stopping producer: [app_name: #{app_name}, " <>
-            "shard_id: #{state.shard_id}, lease_owner: #{state.lease_owner}]"
-        )
+        {:error, :lease_renew_failed} ->
+          Logger.debug(
+            "[kcl_ex] Failed to renew lease, stopping producer: [app_name: #{app_name}, " <>
+              "shard_id: #{state.shard_id}, lease_owner: #{state.lease_owner}]"
+          )
 
-        :ok = Pipeline.stop(app_name, state.shard_id)
-        %{state | lease_holder: false, lease_count_increment_time: current_time()}
+          :ok = Pipeline.stop(app_name, state.shard_id)
+          %{state | lease_holder: false, lease_count_increment_time: current_time()}
 
-      {:error, e} ->
-        Logger.error(
-          "[kcl_ex] Error trying to renew lease for #{state.shard_id} app_name #{app_name}: #{inspect(e)}"
-        )
+        {:error, e} ->
+          Logger.error(
+            "[kcl_ex] Error trying to renew lease for #{state.shard_id} app_name #{app_name}: #{inspect(e)}"
+          )
 
-        state
+          state
+      end
+    else
+      :ok = Pipeline.stop(app_name, state.shard_id)
+      %{state | lease_holder: false, lease_count_increment_time: current_time()}
     end
   end
 
   defp take_lease(_shard_lease, %{app_state_opts: opts, app_name: app_name} = state) do
-    expected = state.lease_count + 1
-
-    Logger.debug(
-      "[kcl_ex] Attempting to take lease: [app_name: #{app_name}, lease_owner: #{state.lease_owner}, shard_id: #{state.shard_id}]"
-    )
-
-    case AppState.take_lease(app_name, state.shard_id, state.lease_owner, state.lease_count, opts) do
-      {:ok, ^expected} ->
-        state = %{
+    if state.can_acquire_lease?.(state.shard_id) do
+      expected = state.lease_count + 1
+
+      Logger.debug(
+        "[kcl_ex] Attempting to take lease: [app_name: #{app_name}, lease_owner: #{state.lease_owner}, shard_id: #{state.shard_id}]"
+      )
+
+      case AppState.take_lease(app_name, state.shard_id, state.lease_owner, state.lease_count, opts) do
+        {:ok, ^expected} ->
+          state = %{
+            state
+            | lease_holder: true,
+              lease_count: expected,
+              lease_count_increment_time: current_time()
+          }
+
+          notify({:lease_taken, state}, state)
+          :ok = Pipeline.start(app_name, state.shard_id)
           state
-          | lease_holder: true,
-            lease_count: expected,
-            lease_count_increment_time: current_time()
-        }
 
-        notify({:lease_taken, state}, state)
-        :ok = Pipeline.start(app_name, state.shard_id)
-        state
-
-      {:error, :lease_take_failed} ->
-        # TODO
-        # :ok = Processor.ensure_halted(state)
-        %{state | lease_holder: false, lease_count_increment_time: current_time()}
+        {:error, :lease_take_failed} ->
+          :ok = Pipeline.stop(app_name, state.shard_id)
```

Comment:
> and any running pipelines are halted.

Thinking out loud here. Are we certain we want to do this here (when lease acquisition fails BUT the current process has been granted allowance to attempt to acquire a lease)? I would think in this case we'd want the current process to stay alive so it can retry? Or, am I misremembering the pipeline hierarchy here?

For the other cases, where we stop the pipeline when allowance is denied, it roughly makes sense to me. This technically means those processes cannot become available to acquire a lease if one is released elsewhere. But, in practice, that doesn't already work and we do full supervision tree restarts to move around leases.

Reply:
Great thought. My read is that the current process _does_ stay alive, but the broadway pipeline (a sibling to the lease process in the supervision tree) is stopped. This seems correct, since if we don't get the lease, then we shouldn't have the lease. Lease acquisition attempts should continue after this point.

So, I think that as-written all pods will be constantly attempting to acquire leases still, but they'll just get short-circuited (and never hit dynamo - side benefit!) if the check function returns false. As the cluster topology changes (e.g. if we add more pods at runtime), leases will shuffle around still to rebalance.